File: solutions/0186_Reverse_Words_in_a_String_II/const_space.cpp
// --Link--
// https://www.lintcode.com/problem/reverse-words-in-a-string-ii/description

class Solution {
public:
    /**
     * @param str: a string
     * @return: return a string
     */
    string reverseWords(string &str) {
        // write your code here
        reverse(str, 0, str.length() - 1);
        int i = 0, j = 0;
        int n = str.length();
        while (i < n) {
            while (i < n && str[i] == ' ') ++i;
            j = i;
            while (j < n && str[j] != ' ') ++j;
            reverse(str, i, j - 1);
            i = j;
        }
        return str;
    }

    void reverse(string &s, int i, int j) {
        while (i < j) {
            char c = s[i];
            s[i] = s[j];
            s[j] = c;
            ++i, --j;
        }
    }
};

File: solutions/0255_Verify_Preorder_Sequence_in_Binary_Search_Tree/const_space.py
# --STAR-- --Link--
# https://www.lintcode.com/problem/verify-preorder-sequence-in-binary-search-tree/description

class Solution:
    """
    @param preorder: List[int]
    @return: return a boolean
    """
    def verifyPreorder(self, preorder):
        # write your code here
        low = -1 << 32
        index = -1 # use preorder as a stack, this is the stack's top
        for v in preorder:
            if v < low:
                return False
            while index >= 0 and preorder[index] < preorder[i]:
                low = preorder[index]
                index -= 1
            preorder[index] = preorder[i]
            index += 1
        return True
